File: packages/sdk-python/tracelane/instrumentations/openrouter.py
"""OpenRouter instrumentation for Tracelane.

OpenRouter exposes an OpenAI-compatible API at https://openrouter.ai/api/v1.
This adapter wraps an OpenAI client that has been pointed at OpenRouter,
overriding the gen_ai.provider.name attribute to "openrouter" and capturing the
resolved model from the response (OpenRouter may reroute to a different
provider model than requested).

Example::

    import openai
    from tracelane.instrumentations.openrouter import instrument_openrouter

    client = openai.OpenAI(
        base_url="https://openrouter.ai/api/v1",
        api_key=os.environ["OPENROUTER_API_KEY"],
    )
    instrument_openrouter(client)
    # chat.completions.create() calls now emit OpenRouter-tagged spans
"""

from __future__ import annotations

import inspect
from typing import Any

from opentelemetry import trace
from opentelemetry.trace import SpanKind, StatusCode

_tracer = trace.get_tracer("tracelane.openrouter", "0.1.0")


def instrument_openrouter(client: Any) -> None:
    """Wrap an OpenAI client pointed at OpenRouter to emit OTel spans.

    Args:
        client: An openai.OpenAI or openai.AsyncOpenAI instance with
                base_url set to https://openrouter.ai/api/v1.

    Note:
        Captures openrouter.route from the response model field, which
        reflects the actual provider model selected by OpenRouter's router.
    """
    original_create = client.chat.completions.create

    if inspect.iscoroutinefunction(original_create):

        async def _async_create(*args: Any, **kwargs: Any) -> Any:
            model = kwargs.get("model", "unknown")
            with _tracer.start_as_current_span(
                "openrouter.chat.completions.create",
                kind=SpanKind.CLIENT,
                attributes={
                    "gen_ai.provider.name": "openrouter",
                    "gen_ai.request.model": model,
                    "llm.model_name": model,
                },
            ) as span:
                try:
                    result = await original_create(*args, **kwargs)
                    _record_response(span, result)
                    span.set_status(StatusCode.OK)
                    return result
                except Exception as exc:
                    span.record_exception(exc)
                    span.set_status(StatusCode.ERROR, str(exc))
                    raise

        client.chat.completions.create = _async_create
    else:

        def _sync_create(*args: Any, **kwargs: Any) -> Any:
            model = kwargs.get("model", "unknown")
            with _tracer.start_as_current_span(
                "openrouter.chat.completions.create",
                kind=SpanKind.CLIENT,
                attributes={
                    "gen_ai.provider.name": "openrouter",
                    "gen_ai.request.model": model,
                    "llm.model_name": model,
                },
            ) as span:
                try:
                    result = original_create(*args, **kwargs)
                    _record_response(span, result)
                    span.set_status(StatusCode.OK)
                    return result
                except Exception as exc:
                    span.record_exception(exc)
                    span.set_status(StatusCode.ERROR, str(exc))
                    raise

        client.chat.completions.create = _sync_create


def _record_response(span: Any, result: Any) -> None:
    usage = getattr(result, "usage", None)
    if usage:
        prompt = getattr(usage, "prompt_tokens", 0) or 0
        completion = getattr(usage, "completion_tokens", 0) or 0
        span.set_attribute("gen_ai.usage.input_tokens", prompt)
        span.set_attribute("gen_ai.usage.output_tokens", completion)
        span.set_attribute("llm.token_count.prompt", prompt)
        span.set_attribute("llm.token_count.completion", completion)

    # OpenRouter returns the actual routed model in response.model
    routed_model = getattr(result, "model", None)
    if routed_model:
        span.set_attribute("gen_ai.response.model", routed_model)
        span.set_attribute("openrouter.route", routed_model)

File: packages/sdk-typescript/src/instrumentations/openrouter.ts
/**
 * OpenRouter instrumentation for Tracelane.
 *
 * OpenRouter uses the OpenAI SDK with a different base URL. This adapter
 * tags spans with gen_ai.provider.name = "openrouter" and captures the routed
 * model from the response (OpenRouter may select a different model
 * than the one requested via its routing logic).
 *
 * @example
 * ```ts
 * import OpenAI from "openai";
 * import { instrumentOpenRouter } from "@tracelanedev/sdk/openrouter";
 *
 * const client = new OpenAI({
 *   baseURL: "https://openrouter.ai/api/v1",
 *   apiKey: process.env.OPENROUTER_API_KEY!,
 * });
 * instrumentOpenRouter(client);
 * ```
 */

import { SpanKind, SpanStatusCode, trace } from "@opentelemetry/api";

const tracer = trace.getTracer("@tracelanedev/sdk-openrouter", "0.1.0");

interface OpenRouterClientLike {
	chat: {
		completions: {
			create: (...args: unknown[]) => Promise<unknown>;
		};
	};
}

/**
 * Instrument an OpenAI client pointed at OpenRouter.
 *
 * @param client - An OpenAI client with baseURL = https://openrouter.ai/api/v1
 */
export function instrumentOpenRouter(client: OpenRouterClientLike): void {
	const originalCreate = client.chat.completions.create.bind(
		client.chat.completions,
	);

	client.chat.completions.create = async (...args: unknown[]) => {
		const request = args[0] as Record<string, unknown>;
		const requestedModel =
			typeof request?.model === "string" ? request.model : "unknown";

		return tracer.startActiveSpan(
			"openrouter.chat.completions.create",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "openrouter",
					"gen_ai.request.model": requestedModel,
					"llm.model_name": requestedModel,
				},
			},
			async (span) => {
				try {
					const result = (await originalCreate(...args)) as Record<
						string,
						unknown
					>;
					const usage = result.usage as Record<string, unknown> | undefined;
					if (usage) {
						span.setAttributes({
							"gen_ai.usage.input_tokens": Number(usage.prompt_tokens ?? 0),
							"gen_ai.usage.output_tokens": Number(
								usage.completion_tokens ?? 0,
							),
						});
					}
					// OpenRouter returns the actual routed model in response.model
					if (typeof result.model === "string") {
						span.setAttributes({
							"gen_ai.response.model": result.model,
							"openrouter.route": result.model,
						});
					}
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}
